### PR TITLE
Implement the CT add-pre-chain-handler

### PR DESCRIPTION
**Pull request by Martin2112**
This is mostly a refactoring and moving code around to break out helpers to make the add chain logic clearer and then a dedup of the code so that add-pre-chain shares everything. Most of the actual new code is tests.

### CODE REVIEW

File: examples/ct/ct_handlers_test.go [hunk 1]
```diff
@@ -357,6 +358,48 @@ func TestAddChain(t *testing.T) {
...
+// Submit a chain with a valid path but using a cert instead of a precert. Should be rejected.
+func TestAddPrecertChainCert(t *testing.T) {
+	km := new(crypto.MockKeyManager)
+	client := new(trillian.MockTrillianLogClient)
+
+	roots := loadCertsIntoPoolOrDie(t, []string{testonly.CACertPEM})
+	reqHandlers := CTRequestHandlers{0x42, roots, client, km, time.Millisecond * 500, fakeTimeSource}
+
+	cert, err := fixchain.CertificateFromPEM(testonly.TestCertPEM)
+	_, ok := err.(x509.NonFatalErrors)
```
Comment (phad) [inline examples/ct/ct_handlers_test.go]: this is a type assertion, right?
so I believe you are checking that CertificateFromPEM's err is nil or an x509.NonFatalError, and will t.Fatal for anything else?

File: examples/ct/ct_handlers_test.go [hunk 2]
```diff
@@ -25,6 +24,7 @@ import (
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/mock"
 	"golang.org/x/net/context"
+	"crypto/sha256"
```
Comment (phad) [inline examples/ct/ct_handlers_test.go]: this import move seems spurious, crypto/sha256 normally comes  in the system imports at the top

File: examples/ct/ct_handlers_test.go [hunk 3]
```diff
@@ -358,12 +372,42 @@ func TestAddChain(t *testing.T) {
...
+	reqHandlers := CTRequestHandlers{0x42, roots, client, km, time.Millisecond * 500, fakeTimeSource}
+
+	cert, err := fixchain.CertificateFromPEM(testonly.PrecertPEMValid)
+	_, ok := err.(x509.NonFatalErrors)
+
+	if err != nil && !ok {
+		t.Fatal(err)
+	}
+
+	pool := NewPEMCertPool()
```
Comment (phad) [inline examples/ct/ct_handlers_test.go]: Does this PEMCertPool struct represent a certificate chain?

Comment (phad) [inline examples/ct/ct_handlers_test.go]: Why the 'pool' name?  Are you memoizing certs once seen and storing a set of unique certs, or something?

File: examples/ct/ct_handlers_test.go [hunk 4]
```diff
@@ -392,12 +436,89 @@ func TestAddPrecertChainCert(t *testing.T) {
...
+		t.Fatal(err)
+	}
+
+	leaves := leafProtosForCert(t, km, pool.RawCertificates(), merkleLeaf)
+
+	client.On("QueueLeaves", mock.MatchedBy(deadlineMatcher), &trillian.QueueLeavesRequest{LogId: 0x42, Leaves: leaves}, mock.Anything /* []grpc.CallOption */).Return(&trillian.QueueLeavesResponse{Status: &trillian.TrillianApiStatus{StatusCode: trillian.TrillianApiStatusCode_OK}}, nil)
+
+	recorder := makeAddPrechainRequest(t, reqHandlers, chain)
+
+	assert.Equal(t, http.StatusOK, recorder.Code, "expected HTTP OK for valid add-chain: %v", recorder.Body)
```
Comment (phad) [inline examples/ct/ct_handlers_test.go]: message: for valid 'add-pre-chain'

File: examples/ct/ct_handlers.go [hunk 5]
```diff
@@ -494,3 +420,64 @@ func marshalLogIDAndSignatureForResponse(sct ct.SignedCertificateTimestamp, km c
...
+	w.Header().Set(contentTypeHeader, contentTypeJSON)
+	jsonData, err := json.Marshal(&resp)
+
+	if err != nil {
+		glog.Warningf("Failed to marshal add-chain resp: %v", resp)
+		sendHttpError(w, http.StatusInternalServerError, err)
+		return err
+	}
+
+	_, err = w.Write(jsonData)
```
Comment (phad) [inline examples/ct/ct_handlers.go]: What is it being thrown away in the first part of the return?

File: examples/ct/ct_handlers_test.go (hunk 1 above)
Comment (Martin2112) [inline examples/ct/ct_handlers_test.go]: Yes, somewhat annoyingly the CT Go code returns NonFatalError when you parse a cert that contains our precert extension.

File: examples/ct/ct_handlers_test.go (hunk 2 above)
Comment (Martin2112) [inline examples/ct/ct_handlers_test.go]: Sometimes IntelliJ inserts them in the wrong place and they don't get fixed up. I've moved it back.

File: examples/ct/ct_handlers_test.go (hunk 3 above)
Comment (Martin2112) [inline examples/ct/ct_handlers_test.go]: It's a container for a bunch of certs. It does store them in the order added so can be used to represent a chain in tests but there's no requirement that the certs added have an relationship. It's a specialized version of CertPool, from the CT Go code so I retained the name.

File: examples/ct/ct_handlers_test.go (hunk 4 above)
Comment (Martin2112) [inline examples/ct/ct_handlers_test.go]: Done.

File: examples/ct/ct_handlers.go (hunk 5 above)
Comment (Martin2112) [inline examples/ct/ct_handlers.go]: The number of bytes written. The docs say writers must return an error if they write less than the data supplied.

File: examples/ct/ct_handlers_test.go (hunk 1 above)
Comment (phad) [inline examples/ct/ct_handlers_test.go]: I see. So, t.Fatal below occurs if (a) there was an err and (b) it wasn't a x509.NonFatalErrors.

So, since you parsed TestCertPEM (which is not a pre-cert) you definitely don't expect an x509.NonFatalErrors err to be returned.

However if TestCertPEM turned out to be a pre-cert the test would continue.

I think what I'm saying is that you'd want t.Fatal called if it turned out that TestCertPEM _was_ a precert.

In other words, the !ok part is superfluous.

File: examples/ct/ct_handlers_test.go (hunk 3 above)
Comment (phad) [inline examples/ct/ct_handlers_test.go]: OK, that's fine, still curious about the name though.  Does it implement pooling behaviour (re-use of objects for example)?

Comment (Martin2112) [inline examples/ct/ct_handlers_test.go]: It originated in golang crypto/x509 but they didn't document why they picked it. I don't think it's ideal because it doesn't do what you suggest.

File: examples/ct/ct_handlers_test.go (hunk 1 above)
Comment (Martin2112) [inline examples/ct/ct_handlers_test.go]: You're right, well spotted. Probably cut and pasted from the precert test. Fixed.

**Comment by Martin2112** [comment]
I think I've addressed all your comments now but it's really hard to be sure with this review UI.

**Comment by phad** [comment]
LGTM for this PR.

I think you might consider tightening up checking of err returned from loading precerts in all the tests.

If I've understood correctly, you expect the X509.NonFatalError return whenever you load a precert, but the tests seem willing to continue running if a nil err is returned.

So code that says
err = LoadCertificateFromPEM(kPreCert)
_, ok := err.(X509.NonFatalErrors)
if err != nil && !ok {  t.Fatal () }

could be tightened up to:
err = LoadCertificateFromPEM(kPreCert)
_, ok := err.(X509.NonFatalErrors)
if err == nil || !ok {  t.Fatal () }

OTOH you might take the view that LoadCertificateFromPEM(kPreCert) ought to just return nil anyway.

**Comment by Martin2112** [comment]
Thanks. As discussed I don't think that error should be returned by CT for something we expect to happen. I'll put a TODO in this file to revisit the tests if this behaviour is fixed upstream.
